### DISCUSSION TITLE
How to get the project files in the whole solution?

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/roslyn#49563] [Q&A]
Hi everyone. I'm new for writing a VS plugin.

Sometimes I want to clean code, but it seems that code fixes don't process some tiny ones, such as the range checks for `int` values like `i is > 10` or `i > 10`. The former one uses an unnecessary keyword `is`. Currently, I want to write code to create a plugin to find all "keyword `is`s" (Indeed, I know that I can use "Find in files" or "find or replaces" to find all keyword `is`s. However, I'd like to create a new plugin to help me find the symbol with the specified kind, specified name or others. In this way I can find all symbols at the specified location quickly and efficiently no matter what I want to find).

However, I'm stuck in the first step. How should I do to get all files in the specified solution?

> I can't decide which one this topic belongs to, Roslyn or Visual Studio itself. If the topic isn't related to Roslyn, I'll be sorry for that; If duplicate, sorry. :(

Thanks a lot. :P

**Comment by PathogenDavid** [comment]
Maybe I'm missing something, but it seems to me that rather than making a Visual Studio extension you should just be making a plain old Roslyn analyzer with a code fix. You won't have to worry about enumerating files and such, the compiler will handle that for you. Here's the official tutorial on writing them: https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix

If you want your code fix to be available at all times, you can deploy them as Visual Studio extensions. (Otherwise you can just distribute it as a NuGet package and install it on projects individually.)

**Reply by kyoyama-kazusa** [reply]
In fact I want to create a WinForm or WPF window to search for those symbols. Your reply can be helpful, but I want to search for a specified symbol type from all files when I input or choose several conditions, rather than reporting a diagnostic result in a single file.

For example, when I want to search them, just choose that command, then VS opens the corresponding window come from the plugin I wrote and create. I can choose some condition to restrict symbol type, such as I just search for nullable suppression operator, then just choose "Nullable suppression operator“ in combo box is okay. Then VS will search for the codes that contain this operator, and gather the result in a result window.

Anyway, thank you for your reply! :D

**Reply by PathogenDavid** [reply]
Ah, that's definitely a more complex situation.

I believe [`DiagnosticSeverity.Hidden`](https://github.com/dotnet/roslyn/blob/2834b74995bb66a7cb19cb09069c17812819afdc/src/Compilers/Core/Portable/Diagnostic/DiagnosticSeverity.cs#L12-L17) is intended to support these sorts of scenarios. Basically you use an analyzer to identify the relevant parts of code and your tool consumes these invisible diagnostics to do whatever you want with them. (You might use multiple diagnostic IDs for different things like one diagnostic with and another for without a nullable suppression operator.)

Unfortunately I couldn't tell you how to get access to them from a Visual Studio extension.

However, based on https://github.com/dotnet/roslyn/issues/26972 I'm under the impression that [Code Cleanup](https://docs.microsoft.com/en-us/visualstudio/ide/code-styles-and-code-cleanup?view=vs-2019#apply-code-styles) works this way under the hood, so you might go exploring there. (I'm pretty sure the code for it is kicking around in the Roslyn repository somewhere.)

Either way, sounds like an interesting project, good luck!

**Reply by kyoyama-kazusa** [reply]
Thanks!
